src/canvas.jl

```diff
@@ -44,35 +44,35 @@ end
 
 # Implementation of the digital differential analyser (DDA)
 function lines!(c::Canvas, x1::Number, y1::Number, x2::Number, y2::Number, color::UserColorType)
-    if (x1 < origin_x(c) && x2 < origin_x(c)) ||
-        (x1 > origin_x(c) + width(c) && x2 > origin_x(c) + width(c))
-        return c
-    end
-    if (y1 < origin_y(c) && y2 < origin_y(c)) ||
-        (y1 > origin_y(c) + height(c) && y2 > origin_y(c) + height(c))
-        return c
-    end
-    toff = x1 - origin_x(c)
-    px1 = toff / width(c) * pixel_width(c)
-    toff = x2 - origin_x(c)
-    px2 = toff / width(c) * pixel_width(c)
-    toff = y1 - origin_y(c)
-    py1 = pixel_height(c) - toff / height(c) * pixel_height(c)
-    toff = y2 - origin_y(c)
-    py2 = pixel_height(c) - toff / height(c) * pixel_height(c)
-    dx = px2 - px1
-    dy = py2 - py1
-    nsteps = abs(dx) > abs(dy) ? abs(dx) : abs(dy)
-    inc_x = dx / nsteps
-    inc_y = dy / nsteps
-    cur_x = px1
-    cur_y = py1
-    fpw = Float64(pixel_width(c))
-    fph = Float64(pixel_height(c))
+    mx = origin_x(c)
+    Mx = origin_x(c) + width(c)
+    (x1 < mx && x2 < mx) || (x1 > Mx && x2 > Mx) && return c
+
+    my = origin_y(c)
+    My = origin_y(c) + height(c)
+    (y1 < my && y2 < my) || (y1 > My && y2 > My) && return c
+
+    x2p(x) = (x - mx) / width(c) * pixel_width(c)
+    y2p(y) = pixel_height(c) - (y - my) / height(c) * pixel_height(c)
+
+    Δx = x2p(x2) - (cur_x = x2p(x1))
+    Δy = y2p(y2) - (cur_y = y2p(y1))
+
+    nsteps = abs(Δx) > abs(Δy) ? abs(Δx) : abs(Δy)
+    nsteps = min(nsteps, typemax(Int32))  # hard limit
+
+    δx = Δx / nsteps
+    δy = Δy / nsteps
+
+    px, Px = extrema([x2p(mx), x2p(Mx)])
+    py, Py = extrema([y2p(my), y2p(My)])
+
     pixel!(c, floor(Int, cur_x), floor(Int, cur_y), color)
-    for i = 1:nsteps
-        cur_x += inc_x
-        cur_y += inc_y
+    for _ = 1:nsteps
```

Comment:
`1000` visually looks good to me.

```suggestion
    # For performance, we draw at most 1000 points for each line
    max_num_iter = 1000
    R = if nsteps > max_num_iter
        range(1, stop=nsteps, length=max_num_iter)
    else
        range(1, stop=nsteps, step=1)
    end
    for _ in R
```

This reduces the drawing time of `lineplot([cos, sin, tan], -π/2, 2π)` from 6.5s to 0.000228 seconds

Reply:
Sure, but what if someone wants to draw a line with more than `1000` pts ? I'd be cautious and leave a rather large upper bound ...

Reply:
I'm not sure... I would bet people won't be sensitive to more than 1000 pts for terminals... since we can't enlarge a terminal..

By calling `@show nsteps` with `lineplot([cos, sin, tan], -π/2, 2π)` I see normal `nsteps` are no-larger than 250.

Reply:
Unfortunately, limiting `nsteps` to even something as large as `100_000` fails the tests.

Reply:
Odd, in my local machine it only breaks the newly added `cos_sin_tan.txt`

Reply:
@t-bltg Anyway, since you're _de factor_ maintainer now I'll leave this to you to decide whether we want this "optimization". I will travel tomorrow so might not be available in the next few days.

Reply:
By using `nsteps = min(nsteps, 100_000)  # hard limit`, and `pkg> test UnicodePlots` I get these errors:
```
Test Summary:     | Pass  Error  Total
tst_canvas.jl     |  116      5    121
  interface       |   84            84
  print and show  |   29      5     34
    BrailleCanvas |    4      1      5
    DensityCanvas |    4      1      5
    BlockCanvas   |    4      1      5
    AsciiCanvas   |    4      1      5
    DotCanvas     |    4      1      5
    HeatmapCanvas |    9             9
  color mixing    |    1             1
  colors          |    1             1
```

Reply:
> since you're de factor maintainer
That's not a reason to do something messy :wink: 
I think I will leave that PR open until we find a better solution ...

Reply:
Ha! `range(1, stop=nsteps, length=100_000)` is different than `range(1, stop=min(nteps, 100_000))`.

Reply:
> Ha! `range(1, stop=nsteps, length=100_000)` is different from `range(1, stop=min(nteps, 100_000))`.

End of the day :sleeping:

Reply:
Ok, so I've limited the `nsteps` to `typemax(Int32)` and `max_num_iter` to `typemax(Int16)` for consistency:
```julia
@time lineplot([sin,tan],-2π,10π)
  0.001015 seconds (534 allocations: 56.203 KiB)
```
@johnnychen94, does it look acceptable ?

Reply:
Thanks for your help !